feat(web-app): add express proxy server to log request

francinelucca opened the pull request:
Closes #426 

- [x] cleanup web-app now unuseful code
- [x] figure out port thing, how does this work?

**Testing/Reviewing**
 
- `npm install`
- run with `npm run dev`, make sure app runs as expected both on `http://localhost:8080` and on `http://localhost:3001`
- run with `npm run dev:secure`, make sure app runs as expected on `https://localhost:8443` (might need to set auth variables -CARBON_IBM_ISV_CLIENT_ID, CARBON_IBM_ISV_CLIENT_SECRET, PASSPORT_STRATEGY_NAME- to be able to get through the login)

francinelucca committed:
wip(web-app): add express proxy serve to log request

francinelucca committed:
feat(web-app): request-proxy, minor additions

jdharvey-ibm commented on services/web-app/package.json:
```diff
@@ -13,14 +13,15 @@
-    "dev:secure": "sh ./scripts/create-local-certificates.sh && RUNNING_SECURELY=1 NODE_TLS_REJECT_UNAUTHORIZED=0 node secure-dev-server.js",
+    "dev": "concurrently \"npm run start-proxy\" \"next dev -p 3001\"",
+    "dev:clean": "npm run clean && npm run dev",
+    "dev:secure": "concurrently  \"sh ./scripts/create-local-certificates.sh && RUNNING_SECURELY=1 NODE_TLS_REJECT_UNAUTHORIZED=0 node secure-dev-server.js\" \"RUNNING_SECURELY=1 npm run start-proxy\"",
     "dev:secure:clean": "npm run clean && npm run dev:secure",
     "lint": "next lint",
     "service:build": "npm run build",
-    "start": "next start",
-    "test": "echo 'No tests defined yet!'"
+    "start": "concurrently \"npm run start-proxy\" \"next start -p 3001\"",
+    "test": "echo 'No tests defined yet!'",
+    "start-proxy": "node proxy-server.js"
```
Let's sync up and re-think all of our scripts here to see if we can simplify them based on our common workflows and use cases.

jdharvey-ibm commented on services/web-app/proxy-server.js:
```diff
@@ -0,0 +1,49 @@
+    performance.clearMarks(req.id)
+    performance.clearMeasures(req.id)
+  },
+  onProxyReq: (_, req) => {
+    req.id = uuidv4()
+    performance.mark(req.id)
+  }
```
let's move onProxyReq to the top so it's chronological.

jdharvey-ibm commented on services/web-app/proxy-server.js:
```diff
@@ -0,0 +1,49 @@
+  getRunMode() === RunMode.Standard || process.env.RUNNING_SECURELY === '1' ? 'https' : 'http'
+const BASE_URL = `${protocol}://localhost:3001`
+
+const apiProxy = createProxyMiddleware('/', {
```
```suggestion
const nextJsProxy = createProxyMiddleware('/', {
```

jdharvey-ibm commented on services/web-app/proxy-server.js:
```diff
@@ -0,0 +1,49 @@
+
+app.use(apiProxy)
+
+const port = process.env.PORT || 3000
```
minor: move this const to the top of the file.

jdharvey-ibm commented on services/web-app/proxy-server.js:
```diff
@@ -0,0 +1,49 @@
+
+const port = process.env.PORT || 3000
+app.listen(port, () => {
+  console.log(`listening on port ${port}`)
```
```suggestion
  logging.info(`listening on port ${port}`)
```

jdharvey-ibm commented on services/web-app/secure-dev-server.js:
```diff
@@ -11,9 +11,9 @@
+const { RunMode, getRunMode } = require('@carbon-platform/api/runtime')
 
-const { PORT } = loadEnvVars({ PORT: '443' })
+const PORT = 3001
```
Any chance for this file that we can remove it and combine it with the new proxy-server file?

https://github.com/http-party/node-http-proxy#using-https

(of particular interest is the [https->http](https://github.com/http-party/node-http-proxy#https---http) section.

These docs seem to indicate that the express app can bear the responsibility for TLS termination, leaving the web-app (proxy target) as-is using http.

Definitely check me on that, but if we can do that, I think it would go a long way toward simplifying the logic.

francinelucca replied:
Done!!

jdharvey-ibm commented on services/web-app/utils/sendLocalRequest.js:
```diff
@@ -37,7 +38,8 @@ export default function sendLocalRequest(
   requestOptions = {}
 ) {
-  const protocol = process.env.RUNNING_SECURELY === '1' ? 'https' : 'http'
+  const protocol =
+    getRunMode() === RunMode.Standard || process.env.RUNNING_SECURELY === '1' ? 'https' : 'http'
```
If we can get TLS termination exclusively in the proxy-server.js file, then internal requests could go back to assuming they're always using http.

francinelucca replied:
Done!! 🎉

francinelucca committed:
Apply suggestions from code review

Co-authored-by: Joe Harvey <51208233+jdharvey-ibm@users.noreply.github.com>

francinelucca committed:
fix(web-app): address code review commennts

francinelucca committed:
fix(web-app): linting issues

jharvey10 committed:
deps(web-app): reinstall new node modules

francinelucca committed:
fix(web-app): cr comments, consolidate secure-server with proxy-server

francinelucca committed:
Merge branch 'main' of github.com:carbon-design-system/carbon-platform into flucca/feat/express-proxy-server

francinelucca committed:
wip(web-app): make dockerfile work with proxy

jharvey10 committed:
fix(web-app): update proxy server

- Add esbuild to pre-package deps to work with proxy server
- Update docker file to run esbuild output
- Fix package.json dev:secure bug
- Refactor proxy server to have less side-effects to facilitate easier unit testing
- Using concatenation for proxy logs for better performance over template literals

francinelucca committed:
fix(web-app): correct web app docs

jdharvey-ibm commented on docs/services-web-app.md:
```diff
@@ -61,19 +51,45 @@ In Standard run mode, the application must run on https for IBMid authentication
+
+`npm run dev` or, alternatively, run `npm run dev:clean` to run from a clean environment
+
+App will run on https://localhost:3000
```
```suggestion
App will run on http://localhost:3000
```

jdharvey-ibm commented on docs/services-web-app.md:
```diff
@@ -61,19 +51,45 @@ In Standard run mode, the application must run on https for IBMid authentication
+`npm run dev` , `npm run start:proxy`
 
-`npm run start-secure`
+App will run on http://localhost:8080, you can still access the unproxied application at
```
```suggestion
The proxied version of the app will run on http://localhost:8080. You can still access the unproxied application at
```

francinelucca committed:
Update docs/services-web-app.md

Co-authored-by: Joe Harvey <51208233+jdharvey-ibm@users.noreply.github.com>

francinelucca committed:
Update docs/services-web-app.md

Co-authored-by: Joe Harvey <51208233+jdharvey-ibm@users.noreply.github.com>

francinelucca committed:
Update services-web-app.md

francinelucca committed:
fix: lint issues
